django/cantusdb_project/main_app/permissions.py

```diff
@@ -0,0 +1,130 @@
+from django.db.models import Q
+
+
+def user_can_edit_chants_in_source(user, source) -> bool:
+    """
+    Checks if the user can edit Chants in a given Source.
+    Used in ChantDetail, ChantList, ChantCreate, ChantDelete, ChantEdit,
+    ChantEditSyllabification, and SourceDetail views.
+    """
+    if user.is_anonymous or (source is None):
+        return False
+
+    source_id = source.id
+    user_is_assigned_to_source: bool = user.sources_user_can_edit.filter(
+        id=source_id
+    ).exists()
+
+    user_is_project_manager: bool = user.groups.filter(name="project manager").exists()
+    user_is_editor: bool = user.groups.filter(name="editor").exists()
+    user_is_contributor: bool = user.groups.filter(name="contributor").exists()
+
+    return (
+        (user_is_project_manager)
+        or (user_is_editor and user_is_assigned_to_source)
+        or (user_is_editor and source.created_by == user)
+        or (user_is_contributor and user_is_assigned_to_source)
+        or (user_is_contributor and source.created_by == user)
+    )
+
+
+def user_is_proofreader(user) -> bool:
+    """
+    Checks if the user is a proofreader.
+    Used in UserSourceListView.
+    """
+    return user.groups.filter(Q(name="project manager") | Q(name="editor")).exists()
+
+
+def user_can_proofread_chants_in_source(user, source_id) -> bool:
+    """
+    Checks if the user can access the proofreading page of a given Source.
+    Used in ChantProofreadView.
+    """
+    if user.is_anonymous:
+        return False
+    user_is_assigned_to_source: bool = user.sources_user_can_edit.filter(
+        id=source_id
+    ).exists()
+
+    user_is_project_manager: bool = user.groups.filter(name="project manager").exists()
+    user_id_editor: bool = user.groups.filter(name="editor").exists()
+
+    if (user_is_project_manager) or (user_id_editor and user_is_assigned_to_source):
+        return True
+    return False
+
+
+def user_can_view_unpublished_source(user, source) -> bool:
+    """
+    Checks if the user can view unpublished Sources or Chants/Sequences belonging
+    to unpublished sources on the site.
+    Used in ChantDetail, SequenceDetail, and SourceDetail views.
+    """
+    user_is_authenticated: bool = user.is_authenticated
+    if (
+        (source is not None)
+        and (source.published is False)
+        and (not user_is_authenticated)
+    ):
+        return False
+    return True
+
+
+def user_can_edit_sequence(user) -> bool:
+    """
+    Checks if the user has permission to edit a Sequence object.
+    Used in SequenceDetail and SequenceEdit views.
+    """
+    return user.groups.filter(name="project manager").exists()
+
+
+def user_can_create_source(user) -> bool:
```

Comment:
This is fine, but maybe `user_can_create_sources` would be better. The current wording slightly suggests that the function should take a `source` argument as well as a `user`